File: data-plane/bindings/python/examples/slimrpc/simple/client_group.py
import asyncio
import logging
import time
from datetime import timedelta

import slim_bindings
from examples.slimrpc.simple.types.example_pb2 import ExampleRequest
from examples.slimrpc.simple.types.example_pb2_slimrpc import TestGroupStub

logger = logging.getLogger(__name__)

# Names of the server instances to broadcast to.
# Start each with: python -m examples.slimrpc.simple.server --instance server1 (or server2)
SERVER_NAMES = [
    slim_bindings.Name("agntcy", "grpc", "server1"),
    slim_bindings.Name("agntcy", "grpc", "server2"),
]


async def run_multicast_unary(stub: TestGroupStub) -> None:
    logger.info("--- multicast unary-unary ---")
    request = ExampleRequest(example_integer=1, example_string="hello")
    async for context, resp in stub.ExampleUnaryUnary(
        request, timeout=timedelta(seconds=5)
    ):
        logger.info(f"  [{context}] {resp}")


async def run_multicast_unary_stream(stub: TestGroupStub) -> None:
    logger.info("--- multicast unary-stream ---")
    request = ExampleRequest(example_integer=1, example_string="hello")
    async for context, resp in stub.ExampleUnaryStream(
        request, timeout=timedelta(seconds=20)
    ):
        logger.info(f"{time.time()}")
        logger.info(f"  [{context}] {resp}")


async def stream_requests():
    for i in range(3):
        yield ExampleRequest(example_integer=i, example_string=f"item {i}")


async def run_multicast_stream_unary(stub: TestGroupStub) -> None:
    logger.info("--- multicast stream-unary ---")
    async for context, resp in stub.ExampleStreamUnary(
        stream_requests(), timeout=timedelta(seconds=5)
    ):
        logger.info(f"  [{context}] {resp}")


async def run_multicast_stream_stream(stub: TestGroupStub) -> None:
    logger.info("--- multicast stream-stream ---")
    async for context, resp in stub.ExampleStreamStream(
        stream_requests(), timeout=timedelta(seconds=5)
    ):
        logger.info(f"  [{context}] {resp}")


async def amain() -> None:
    slim_bindings.uniffi_set_event_loop(asyncio.get_running_loop())  # type: ignore[arg-type]

    tracing_config = slim_bindings.new_tracing_config()
    runtime_config = slim_bindings.new_runtime_config()
    service_config = slim_bindings.new_service_config()

    tracing_config.log_level = "info"

    slim_bindings.initialize_with_configs(
        tracing_config=tracing_config,
        runtime_config=runtime_config,
        service_config=[service_config],
    )

    service = slim_bindings.get_global_service()

    local_name = slim_bindings.Name("agntcy", "grpc", "client")

    client_config = slim_bindings.new_insecure_client_config("http://localhost:46357")
    conn_id = await service.connect_async(client_config)

    local_app = service.create_app_with_secret(
        local_name, "my_shared_secret_for_testing_purposes_only"
    )
    await local_app.subscribe_async(local_name, conn_id)

    # Group channel targeting all server instances
    channel = slim_bindings.Channel.new_group_with_connection(
        local_app, SERVER_NAMES, conn_id
    )

    stub = TestGroupStub(channel)

    try:
        # await run_multicast_unary(stub)
        # await run_multicast_unary_stream(stub)
        # await run_multicast_stream_unary(stub)
        await run_multicast_stream_stream(stub)
    except slim_bindings.RpcError as e:
        logger.error(f"RPC error: {e}")

    # Close the channel
    await channel.close_async(timeout=None)


def main() -> None:
    logging.basicConfig(level=logging.INFO)
    try:
        asyncio.run(amain())
    except KeyboardInterrupt:
        print("Client interrupted by user.")

File: data-plane/bindings/go/examples/slimrpc/simple/cmd/client_group/client_group.go
//go:build slim_multicast

// Copyright AGNTCY Contributors (https://github.com/agntcy)
// SPDX-License-Identifier: Apache-2.0

package main

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"

	slim_bindings "github.com/agntcy/slim-bindings-go"

	pb "github.com/agntcy/slim/bindings/go/examples/slimrpc/simple/types"
)

const (
	slimAddr     = "http://localhost:46357"
	sharedSecret = "my_shared_secret_for_testing_purposes_only"
)

func runMulticastUnary(client pb.TestGroupClient) {
	log.Println("=== Multicast Unary-Unary ===")
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	stream, err := client.ExampleUnaryUnary(ctx, &pb.ExampleRequest{ExampleInteger: 1, ExampleString: "hello"})
	if err != nil {
		log.Printf("ExampleUnaryUnary failed: %v", err)
		return
	}
	for {
		item, err := stream.Recv()
		if err != nil {
			log.Printf("Recv error: %v", err)
			return
		}
		if item == nil {
			break
		}
		log.Printf("  [%s] %+v", item.Context, item.Value)
	}
}

func runMulticastUnaryStream(client pb.TestGroupClient) {
	log.Println("\n=== Multicast Unary-Stream ===")
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	stream, err := client.ExampleUnaryStream(ctx, &pb.ExampleRequest{ExampleInteger: 1, ExampleString: "hello"})
	if err != nil {
		log.Printf("ExampleUnaryStream failed: %v", err)
		return
	}
	for {
		item, err := stream.Recv()
		if err != nil {
			log.Printf("Recv error: %v", err)
			return
		}
		if item == nil {
			break
		}
		log.Printf("  [%s] %+v", item.Context, item.Value)
	}
}

func runMulticastStreamUnary(client pb.TestGroupClient) {
	log.Println("\n=== Multicast Stream-Unary ===")
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	stream, err := client.ExampleStreamUnary(ctx)
	if err != nil {
		log.Printf("ExampleStreamUnary failed: %v", err)
		return
	}

	for i := int64(0); i < 3; i++ {
		if err := stream.Send(&pb.ExampleRequest{ExampleInteger: i, ExampleString: fmt.Sprintf("item %d", i)}); err != nil {
			log.Printf("Send failed: %v", err)
			return
		}
	}
	if err := stream.CloseSend(); err != nil {
		log.Printf("CloseSend failed: %v", err)
		return
	}

	for {
		item, err := stream.Recv()
		if err != nil {
			log.Printf("Recv error: %v", err)
			return
		}
		if item == nil {
			break
		}
		log.Printf("  [%s] %+v", item.Context, item.Value)
	}
}

func runMulticastStreamStream(client pb.TestGroupClient) {
	log.Println("\n=== Multicast Stream-Stream ===")
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	stream, err := client.ExampleStreamStream(ctx)
	if err != nil {
		log.Printf("ExampleStreamStream failed: %v", err)
		return
	}

	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		for i := int64(0); i < 3; i++ {
			if err := stream.Send(&pb.ExampleRequest{ExampleInteger: i, ExampleString: fmt.Sprintf("item %d", i)}); err != nil {
				log.Printf("Send failed: %v", err)
				return
			}
		}
		if err := stream.CloseSend(); err != nil {
			log.Printf("CloseSend failed: %v", err)
		}
	}()

	for {
		item, err := stream.Recv()
		if err != nil {
			log.Printf("Recv error: %v", err)
			wg.Wait()
			return
		}
		if item == nil {
			wg.Wait()
			break
		}
		log.Printf("  [%s] %+v", item.Context, item.Value)
	}
}

func main() {
	slim_bindings.InitializeWithDefaults()

	service := slim_bindings.GetGlobalService()

	localName := slim_bindings.NewName("agntcy", "grpc", "client")
	serverNames := []*slim_bindings.Name{
		slim_bindings.NewName("agntcy", "grpc", "server1"),
		slim_bindings.NewName("agntcy", "grpc", "server2"),
	}

	app, err := service.CreateAppWithSecret(localName, sharedSecret)
	if err != nil {
		log.Fatalf("Failed to create app: %v", err)
	}

	clientConfig := slim_bindings.NewInsecureClientConfig(slimAddr)
	connId, err := service.Connect(clientConfig)
	if err != nil {
		log.Fatalf("Failed to connect: %v", err)
	}

	if err := app.Subscribe(localName, &connId); err != nil {
		log.Fatalf("Failed to subscribe: %v", err)
	}

	// Group channel targeting all server instances
	channel := slim_bindings.ChannelNewGroupWithConnection(app, serverNames, &connId)

	client := pb.NewTestGroupClient(channel)

	runMulticastUnary(client)
	runMulticastUnaryStream(client)
	runMulticastStreamUnary(client)
	runMulticastStreamStream(client)

	time.Sleep(1 * time.Second)
}
